specs/core/0_beacon-chain.md

```diff
@@ -1615,74 +1569,55 @@ def process_crosslinks(state: BeaconState) -> None:
-            # Inclusion speed bonus
-            earliest_attestation = get_earliest_attestation(state, state.previous_epoch_attestations, index)
-            inclusion_delay = earliest_attestation.inclusion_slot - earliest_attestation.data.slot
-            rewards[index] += base_reward * MIN_ATTESTATION_INCLUSION_DELAY // inclusion_delay
-        else:
-            penalties[index] += base_reward
-        # Expected FFG target
-        if index in get_unslashed_attesting_indices(state, boundary_attestations):
-            rewards[index] += base_reward * boundary_attesting_balance // total_balance
-        else:
-            penalties[index] += get_inactivity_penalty(state, index, epochs_since_finality)
-        # Expected head
-        if index in get_unslashed_attesting_indices(state, matching_head_attestations):
-            rewards[index] += base_reward * matching_head_balance // total_balance
-        else:
-            penalties[index] += base_reward
-        # Take away max rewards if we're not finalizing
-        if epochs_since_finality > 4:
-            penalties[index] += base_reward * 4
+
+    # Micro-incentives for matching FFG source, FFG target, and head
+    matching_source_attestations = get_matching_source_attestations(state, previous_epoch)
+    matching_target_attestations = get_matching_target_attestations(state, previous_epoch)
+    matching_head_attestations = get_matching_head_attestations(state, previous_epoch)
+    for attestations in (matching_source_attestations, matching_target_attestations, matching_head_attestations):
+        unslashed_attesting_indices = get_unslashed_attesting_indices(state, attestations)
+        attesting_balance = get_attesting_balance(state, attestations)
+        for index in eligible_validator_indices:
+            if index in unslashed_attesting_indices:
+                rewards[index] += get_base_reward(state, index) * attesting_balance // total_balance
+            else:
+                penalties[index] += get_base_reward(state, index)
+
+    # Proposer and inclusion delay micro-rewards
+    for index in get_unslashed_attesting_indices(state, matching_source_attestations):
+        earliest_attestation = get_earliest_attestation(state, matching_source_attestations, index)
+        rewards[earliest_attestation.proposer_index] += get_base_reward(state, index) // PROPOSER_REWARD_QUOTIENT
+        inclusion_delay = earliest_attestation.inclusion_slot - earliest_attestation.data.slot
+        rewards[index] += get_base_reward(state, index) * MIN_ATTESTATION_INCLUSION_DELAY // inclusion_delay
+
+    # Inactivity penalty
+    finality_delay = previous_epoch - state.finalized_epoch
+    if finality_delay > MIN_EPOCHS_TO_INACTIVITY_PENALTY:
+        matching_target_attesting_indices = get_unslashed_attesting_indices(state, matching_target_attestations)
+        for index in eligible_validator_indices:
+            penalties[index] += BASE_REWARDS_PER_EPOCH * get_base_reward(state, index)
```

Comment:
Could you elaborate `BASE_REWARDS_PER_EPOCH:= 5`? Does that mean:
1. attesting expected source
2. attesting expected target
3. attesting expected head
4. attesting shard block (in `get_crosslink_deltas`)
5. attestation inclusion

Reply:
1. expected source
2. expected target
3. expected head
4. inclusion delay reward
5. expected crosslink data root